doc/developer/design/20240502_deprecate_subsources.md

```diff
@@ -0,0 +1,228 @@
+
+#### Why not just rename `SUBSOURCE` to `TABLE`?
+
+This leaves the impedance mismatch in place where users have to `SELECT` from
+the source if they're using e.g. Kafka, but have to find the right `TABLE` to
+`SELECT` from for PG sources.
+
+### SOURCES should contain their progress information
+
+The other type of `SUBSOURCE` Materialize supports is known as the
+remap/progress collection. While most users don't access this data, it's an
+important debugging tool.
+
+I propose that the progress collection should be moved "into" the source, i.e.
+querying a source should return the what-is-today-the-progress-collection's
+data.
+
+### New API
+
+The API to support this is similar to the existing `CREATE SOURCE` API. However,
+`ALTER SOURCE...ADD SUBSOURCE` will disappear and be replaced with a flavor of
+`CREATE TABLE`.
+
+(h/t @benesch for this)
+
+---
+
+Create a Kafka source that reads from a topic named `t`:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR ALL TABLES
+```
+
+`kafka_src` is the progress relation for the source, and a table named `t` is
+automatically created in the same schema as `kafka_src`.
+
+Note that we could also probably make `FOR ALL TABLES` implied for Kafka
+sources.
+
+---
+
+Create a Kafka source that reads from a topic named `t` with a custom table
+name:
+
+```sql
+CREATE SOURCE kafka_src
+FROM KAFKA CONNECTION kafka_conn (TOPIC 't')
+FOR TABLES (t AS my_t)
+```
+
+`kafka_src` is again the progress relation for the source. The data is in a
+table named `my_t`. You could also imagine more complex scenarios where you can
+output the same topic with multiple schemas. What that ends up looking like is
+under discussion.
+
+---
+
+Create a PostgreSQL source:
+
+```sql
+CREATE SOURCE postgres_src
+FROM POSTGRES CONNECTION postgres_conn (PUBLICATION p)
+FOR ALL TABLES
+```
+
+`postgres_src` is the progress relation; there is no separate
+`postgres_src_progress` relation created. Otherwise this command works the same
+way that it does today.
+
+---
+
+Add a new table to a PostgreSQL source:
+
+```sql
+CREATE TABLE my_new_table FROM SOURCE postgres_src (upstream_schema.upstream_table);
+```
+
+This would replace the `ALTER SOURCE ... ADD {SUBSOURCE|TABLE}` command.
+
+---
+
+Add a new table to a Kafka source:
+
+```sql
+CREATE TABLE kafka_copy FROM SOURCE kafka_src ("t");
+```
+
+At first this looks weird. Why would you ever want two copies of a Kafka source?
+It turns out this lets you incorporate upstream schema changes without losing
+reclocking information! This would get us reusable reclocking
+(https://github.com/MaterializeInc/materialize/issues/17021), without needing to
+introduce a new `SOURCE PROGRESS` concept.
+
+Note that support for this is blocked on at least
+https://github.com/MaterializeInc/materialize/issues/24843, and also additional
+work to support multi-output sources in Kafka https://github.com/MaterializeInc/materialize/issues/26994.
+
+#### System tables
```

Comment:
This will have a pretty significant impact in the console, so it'd be good to get @RobinClowers' 👀. More broadly, could you expand on how this affects existing core system catalog tables like `mz_sources`, `mz_source_status_history`, or `mz_source_statistics`?

Reply:
Thanks for looping me in Marta, I have the same questions about status and statistics, but the change outlined here is pretty straightforward for the console.